### PR TITLE
[docs] Update guidance on how to implement and register new models

### PR DESCRIPTION
**Pull request by linfeng-yuan**
… vLLM-Ascend

### What this PR does / why we need it?

Update guidance on implementing and registering models for users. [Only documentation]

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Not involve codes and functions.

### CODE REVIEW

**Comment by wangxiyuan** [comment]
mv this doc to developer guide

**Comment by Yikun** [comment]
@shen-shanshan could help review.

File: docs/source/developer_guide/model_registration.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
+# Integrating New Models into vLLM-Ascend
+
+This guide demonstrates how to integrate novel or customized models into vLLM-Ascend. For foundational concepts, it is highly recommended to refer to:
+[Adding a New Model - vLLM Documentation](https://docs.vllm.ai/en/stable/contributing/model/)
+
+### 1. Implementing Models Using PyTorch and Ascend Extension for PyTorch
```
Comment (shen-shanshan) [inline docs/source/developer_guide/model_registration.md:6]: Please fix the CI failure:

`WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]`

Use `##` before `###`.

File: docs/source/developer_guide/model_registration.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+(1) Place model files in [vllm_ascend/models/](https://github.com/vllm-project/vllm-ascend/tree/main/vllm_ascend/models) directory
+
+(2) Standard module structure for decoder-only LLMs (please checkout vllm's implementations for other kinds of model):
+   
+   - `*ModelForCausalLM` (top-level wrapper)
+   - `*Model` (main architecture)
+   - `*DecoderLayer` (transformer block)
+   - `*Attention` & `*MLP` (specific computation unit)
+     `*` denotes your model's unique identifier
```
Comment (shen-shanshan) [inline docs/source/developer_guide/model_registration.md:28]: It's better to add this line into a `NOTE` block.

File: docs/source/developer_guide/model_registration.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+- Required interfaces:
+
+  | Module Type          | Required Methods                          |
+  | :------------------- | :---------------------------------------- |
+  | `*ModelForCausalLM`  | `get_input_embeddings`, `compute_logits`, `load_weights` |
+  | `*Model`             | `get_input_embeddings`, `load_weights`    |
+     
+     
+(4) **Attention Backend Integration**:
+   Import attention via `from vllm.attention import Attention` can automatically leverage vLLM-Ascend's attention backend routing (see: `get_attn_backend_cls()` in [vllm_ascend/platform.py](https://github.com/vllm-project/vllm-ascend/blob/main/vllm_ascend/platform.py))
```
Comment (shen-shanshan) [inline docs/source/developer_guide/model_registration.md:41]: Please make sure adding a `.` at the end of each sentences.

File: docs/source/developer_guide/model_registration.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+(5) **Tensor Parallelism**:
+   Use vLLM's parallel layers (`ColumnParallelLinear`, `VocabParallelEmbedding`, etc.), but note Ascend-specific customizations implemented in [vllm_ascend/ops/](https://github.com/vllm-project/vllm-ascend/tree/main/vllm_ascend/ops) directory (RMSNorm, VocabParallelEmbedding, etc.).
```
Comment (shen-shanshan) [inline docs/source/developer_guide/model_registration.md:44]: I think this is a little bit confused, please make it clear of when to use vLLM's parallel layers and when to use ascend-specific customizations.

File: docs/source/developer_guide/model_registration.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            attn_metadata, 
+            intermediate_tensors,
+            inputs_embeds
+        )
+        return hidden_states
+```
+
+For a complete implementation reference, see: [vllm_ascend/models/deepseek_v2.py](https://github.com/vllm-project/vllm-ascend/blob/main/vllm_ascend/models/deepseek_v2.py)
+
+### 2. Registering Custom Models as Out-of-Tree Plugins in vLLM
```
Comment (shen-shanshan) [inline docs/source/developer_guide/model_registration.md:169]: Registering Custom Out-of-Tree Models using ModelRegistry Plugins in vLLM

File: docs/source/developer_guide/model_registration.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+  ],
+}
+```
+
+```json
+{
+  "architectures": [
+    "DeepseekV2ForCausalLM"
+  ],
+}
```
Comment (shen-shanshan) [inline docs/source/developer_guide/model_registration.md:216]: It's better to add this line into a `NOTE` block.

File: docs/source/developer_guide/model_registration.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+```
+Model architecture DeepseekV2ForCausalLM is already registered, and will be overwritten by the new model class vllm_ascend/models/deepseek_v2:CustomDeepseekV2ForCausalLM.
+```
+
+#### 3.2 Registering New Model Architecture
+
+If you're registering a novel model architecture not present in vLLM (creating a completely new class), current logs won't provide explicit confirmation by default. It's recommended to add the following logging statement at the end of the `register_model` method in [vllm/models\_executor/models/registry.py](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/registry.py) :
+
+```
+logger.warning(f"model_arch: {model_arch} has been registered here!")
```
Comment (shen-shanshan) [inline docs/source/developer_guide/model_registration.md:234]: Why use `warning`? I think use `info` may be better.

**Comment by shen-shanshan** [comment]
@linfeng-yuan I think it's better to add more contents like:

1. The mapping from `torch` ops to `torch_npu` ops, you can add the frequently used interfaces and offer link to more details like [this](https://www.hiascend.com/document/detail/zh/Pytorch/700/apiref/apilist/ptaoplist_000005.html).
2. After adding a new model, we need to do functional test, benchmark and accuracy test, you can offer links of how to do these things.
3. After adding a new model, we should also update our [model support doc](https://vllm-ascend.readthedocs.io/en/latest/user_guide/supported_models.html).
4. Offer link of how to add a multi-modal model.

**Comment by wangxiyuan** [comment]
developer_guide/index should be updated as well

**Comment by wangxiyuan** [comment]
@shen-shanshan it's good to create a new one with co-author, if the path is not updated for a long time.

**Comment by shen-shanshan** [comment]
> @shen-shanshan it's good to create a new one with co-author, if the path is not updated for a long time.

OK.
